File: src/main/java/com/sfh/pokeRogueBot/model/enums/PokeType.java
package com.sfh.pokeRogueBot.model.enums;

import lombok.Getter;

@Getter
public enum PokeType {
    NORMAL, //0
    FIGHTING, //1
    FLYING, //2
    POISON, //3
    GROUND, //4
    ROCK, //5
    BUG, //6
    GHOST, //7
    STEEL, //8
    FIRE, //9
    WATER, //10
    GRASS, //11
    ELECTRIC, //12
    PSYCHIC, //13
    ICE, //14
    DRAGON, //15
    DARK, //16
    FAIRY, //17
    STELLAR; //18

    public static final float[] DAMAGE_MULTIPLIER = new float[]{
            0f, 0.125f, 0.25f, 0.5f, 1f, 2f, 4f, 8f
    };

    public static float getTypeDamageMultiplier(PokeType attackType, PokeType defType) {
        if (attackType == null || defType == null) {
            throw new IllegalArgumentException("Unknown or null type");
        }

        switch (defType) {
            case NORMAL:
                switch (attackType) {
                    case FIGHTING:
                        return 2;
                    case NORMAL:
                    case FLYING:
                    case POISON:
                    case GROUND:
                    case ROCK:
                    case BUG:
                    case STEEL:
                    case FIRE:
                    case WATER:
                    case GRASS:
                    case ELECTRIC:
                    case PSYCHIC:
                    case ICE:
                    case DRAGON:
                    case DARK:
                    case FAIRY:
                        return 1;
                    case GHOST:
                    default:
                        return 0;
                }
            case FIGHTING:
                switch (attackType) {
                    case FLYING:
                    case PSYCHIC:
                    case FAIRY:
                        return 2;
                    case NORMAL:
                    case FIGHTING:
                    case POISON:
                    case GROUND:
                    case GHOST:
                    case STEEL:
                    case FIRE:
                    case WATER:
                    case GRASS:
                    case ELECTRIC:
                    case ICE:
                    case DRAGON:
                        return 1;
                    case ROCK:
                    case BUG:
                    case DARK:
                        return 0.5f;
                    default:
                        return 0;
                }
            case FLYING:
                switch (attackType) {
                    case ROCK:
                    case ELECTRIC:
                    case ICE:
                        return 2;
                    case NORMAL:
                    case FLYING:
                    case POISON:
                    case GHOST:
                    case STEEL:
                    case FIRE:
                    case WATER:
                    case PSYCHIC:
                    case DRAGON:
                    case DARK:
                    case FAIRY:
                        return 1;
                    case FIGHTING:
                    case BUG:
                    case GRASS:
                        return 0.5f;
                    case GROUND:
                    default:
                        return 0;
                }
            case POISON:
                switch (attackType) {
                    case GROUND:
                    case PSYCHIC:
                        return 2;
                    case NORMAL:
                    case FLYING:
                    case ROCK:
                    case GHOST:
                    case STEEL:
                    case FIRE:
                    case WATER:
                    case ELECTRIC:
                    case ICE:
                    case DRAGON:
                    case DARK:
                        return 1;
                    case FIGHTING:
                    case POISON:
                    case BUG:
                    case GRASS:
                    case FAIRY:
                        return 0.5f;
                    default:
                        return 0;
                }
            case GROUND:
                switch (attackType) {
                    case WATER:
                    case GRASS:
                    case ICE:
                        return 2;
                    case NORMAL:
                    case FIGHTING:
                    case FLYING:
                    case GROUND:
                    case BUG:
                    case GHOST:
                    case STEEL:
                    case FIRE:
                    case PSYCHIC:
                    case DRAGON:
                    case DARK:
                    case FAIRY:
                        return 1;
                    case POISON:
                    case ROCK:
                        return 0.5f;
                    case ELECTRIC:
                    default:
                        return 0;
                }
            case ROCK:
                switch (attackType) {
                    case FIGHTING:
                    case GROUND:
                    case STEEL:
                    case WATER:
                    case GRASS:
                        return 2;
                    case ROCK:
                    case BUG:
                    case GHOST:
                    case ELECTRIC:
                    case PSYCHIC:
                    case ICE:
                    case DRAGON:
                    case DARK:
                    case FAIRY:
                        return 1;
                    case NORMAL:
                    case FLYING:
                    case POISON:
                    case FIRE:
                        return 0.5f;
                    default:
                        return 0;
                }
            case BUG:
                switch (attackType) {
                    case FLYING:
                    case ROCK:
                    case FIRE:
                        return 2;
                    case NORMAL:
                    case POISON:
                    case BUG:
                    case GHOST:
                    case STEEL:
                    case WATER:
                    case ELECTRIC:
                    case PSYCHIC:
                    case ICE:
                    case DRAGON:
                    case DARK:
                    case FAIRY:
                        return 1;
                    case FIGHTING:
                    case GROUND:
                    case GRASS:
                        return 0.5f;
                    default:
                        return 0;
                }
            case GHOST:
                switch (attackType) {
                    case GHOST:
                    case DARK:
                        return 2;
                    case FLYING:
                    case GROUND:
                    case ROCK:
                    case STEEL:
                    case FIRE:
                    case WATER:
                    case GRASS:
                    case ELECTRIC:
                    case PSYCHIC:
                    case ICE:
                    case DRAGON:
                    case FAIRY:
                        return 1;
                    case POISON:
                    case BUG:
                        return 0.5f;
                    case NORMAL:
                    case FIGHTING:
                    default:
                        return 0;
                }
            case STEEL:
                switch (attackType) {
                    case FIGHTING:
                    case GROUND:
                    case FIRE:
                        return 2;
                    case GHOST:
                    case WATER:
                    case ELECTRIC:
                    case DARK:
                        return 1;
                    case NORMAL:
                    case FLYING:
                    case ROCK:
                    case BUG:
                    case STEEL:
                    case GRASS:
                    case PSYCHIC:
                    case ICE:
                    case DRAGON:
                    case FAIRY:
                        return 0.5f;
                    case POISON:
                    default:
                        return 0;
                }
            case FIRE:
                switch (attackType) {
                    case GROUND:
                    case ROCK:
                    case WATER:
                        return 2;
                    case NORMAL:
                    case FIGHTING:
                    case FLYING:
                    case POISON:
                    case GHOST:
                    case ELECTRIC:
                    case PSYCHIC:
                    case DRAGON:
                    case DARK:
                        return 1;
                    case BUG:
                    case STEEL:
                    case FIRE:
                    case GRASS:
                    case ICE:
                    case FAIRY:
                        return 0.5f;
                    default:
                        return 0;
                }
            case WATER:
                switch (attackType) {
                    case GRASS:
                    case ELECTRIC:
                        return 2;
                    case NORMAL:
                    case FIGHTING:
                    case FLYING:
                    case POISON:
                    case GROUND:
                    case ROCK:
                    case BUG:
                    case GHOST:
                    case PSYCHIC:
                    case DRAGON:
                    case DARK:
                    case FAIRY:
                        return 1;
                    case STEEL:
                    case FIRE:
                    case WATER:
                    case ICE:
                        return 0.5f;
                    default:
                        return 0;
                }
            case GRASS:
                switch (attackType) {
                    case FLYING:
                    case POISON:
                    case BUG:
                    case FIRE:
                    case ICE:
                        return 2;
                    case NORMAL:
                    case FIGHTING:
                    case ROCK:
                    case GHOST:
                    case STEEL:
                    case PSYCHIC:
                    case DRAGON:
                    case DARK:
                    case FAIRY:
                        return 1;
                    case GROUND:
                    case WATER:
                    case GRASS:
                    case ELECTRIC:
                        return 0.5f;
                    default:
                        return 0;
                }
            case ELECTRIC:
                switch (attackType) {
                    case GROUND:
                        return 2;
                    case NORMAL:
                    case FIGHTING:
                    case POISON:
                    case ROCK:
                    case BUG:
                    case GHOST:
                    case FIRE:
                    case WATER:
                    case GRASS:
                    case PSYCHIC:
                    case ICE:
                    case DRAGON:
                    case DARK:
                    case FAIRY:
                        return 1;
                    case FLYING:
                    case STEEL:
                    case ELECTRIC:
                        return 0.5f;
                    default:
                        return 0;
                }
            case PSYCHIC:
                switch (attackType) {
                    case BUG:
                    case GHOST:
                    case DARK:
                        return 2;
                    case NORMAL:
                    case FLYING:
                    case POISON:
                    case GROUND:
                    case ROCK:
                    case STEEL:
                    case FIRE:
                    case WATER:
                    case GRASS:
                    case ELECTRIC:
                    case ICE:
                    case DRAGON:
                    case FAIRY:
                        return 1;
                    case FIGHTING:
                    case PSYCHIC:
                        return 0.5f;
                    default:
                        return 0;
                }
            case ICE:
                switch (attackType) {
                    case FIGHTING:
                    case ROCK:
                    case STEEL:
                    case FIRE:
                        return 2;
                    case NORMAL:
                    case FLYING:
                    case POISON:
                    case GROUND:
                    case BUG:
                    case GHOST:
                    case WATER:
                    case GRASS:
                    case ELECTRIC:
                    case PSYCHIC:
                    case DRAGON:
                    case DARK:
                    case FAIRY:
                        return 1;
                    case ICE:
                        return 0.5f;
                    default:
                        return 0;
                }
            case DRAGON:
                switch (attackType) {
                    case ICE:
                    case DRAGON:
                    case FAIRY:
                        return 2;
                    case NORMAL:
                    case FIGHTING:
                    case FLYING:
                    case POISON:
                    case GROUND:
                    case ROCK:
                    case BUG:
                    case GHOST:
                    case STEEL:
                    case PSYCHIC:
                    case DARK:
                        return 1;
                    case FIRE:
                    case WATER:
                    case GRASS:
                    case ELECTRIC:
                        return 0.5f;
                    default:
                        return 0;
                }
            case DARK:
                switch (attackType) {
                    case FIGHTING:
                    case BUG:
                    case FAIRY:
                        return 2;
                    case NORMAL:
                    case FLYING:
                    case POISON:
                    case GROUND:
                    case ROCK:
                    case STEEL:
                    case FIRE:
                    case WATER:
                    case GRASS:
                    case ELECTRIC:
                    case ICE:
                    case DRAGON:
                        return 1;
                    case GHOST:
                    case DARK:
                        return 0.5f;
                    case PSYCHIC:
                    default:
                        return 0;
                }
            case FAIRY:
                switch (attackType) {
                    case POISON:
                    case STEEL:
                        return 2;
                    case NORMAL:
                    case FLYING:
                    case GROUND:
                    case ROCK:
                    case GHOST:
                    case FIRE:
                    case WATER:
                    case GRASS:
                    case ELECTRIC:
                    case PSYCHIC:
                    case ICE:
                    case FAIRY:
                        return 1;
                    case FIGHTING:
                    case BUG:
                    case DARK:
                        return 0.5f;
                    case DRAGON:
                    default:
                        return 0;
                }
            case STELLAR:
                return 1;
        }

        return 1; //fallback
    }

    public static String getAttackTypeBoosterItemName(PokeType type) {
        switch (type) {
            case NORMAL:
                return "Silk Scarf";
            case FIGHTING:
                return "Black Belt";
            case FLYING:
                return "Sharp Beak";
            case POISON:
                return "Poison Barb";
            case GROUND:
                return "Soft Sand";
            case ROCK:
                return "Hard Stone";
            case BUG:
                return "Silver Powder";
            case GHOST:
                return "Spell Tag";
            case STEEL:
                return "Metal Coat";
            case FIRE:
                return "Charcoal";
            case WATER:
                return "Mystic Water";
            case GRASS:
                return "Miracle Seed";
            case ELECTRIC:
                return "Magnet";
            case PSYCHIC:
                return "Twisted Spoon";
            case ICE:
                return "Never-Melt Ice";
            case DRAGON:
                return "Dragon Fang";
            case DARK:
                return "Black Glasses";
            case FAIRY:
                return "Fairy Feather";
        }

        return null; //fallback
    }
}

File: src/main/kotlin/com/sfh/pokeRogueBot/model/enums/PokeType.kt
package com.sfh.pokeRogueBot.model.enums

enum class PokeType {
    NORMAL, //0
    FIGHTING, //1
    FLYING, //2
    POISON, //3
    GROUND, //4
    ROCK, //5
    BUG, //6
    GHOST, //7
    STEEL, //8
    FIRE, //9
    WATER, //10
    GRASS, //11
    ELECTRIC, //12
    PSYCHIC, //13
    ICE, //14
    DRAGON, //15
    DARK, //16
    FAIRY, //17
    STELLAR; //18

    companion object {
        val DAMAGE_MULTIPLIER = floatArrayOf(
            0f, 0.125f, 0.25f, 0.5f, 1f, 2f, 4f, 8f
        )

        fun getTypeDamageMultiplier(attackType: PokeType?, defType: PokeType?): Float {
            if (attackType == null || defType == null) {
                throw IllegalArgumentException("Unknown or null type")
            }

            return when (defType) {
                NORMAL -> when (attackType) {
                    FIGHTING -> 2f
                    NORMAL, FLYING, POISON, GROUND, ROCK, BUG, STEEL, FIRE, WATER, GRASS, ELECTRIC, PSYCHIC, ICE, DRAGON, DARK, FAIRY -> 1f
                    GHOST -> 0f
                    else -> 0f
                }

                FIGHTING -> when (attackType) {
                    FLYING, PSYCHIC, FAIRY -> 2f
                    NORMAL, FIGHTING, POISON, GROUND, GHOST, STEEL, FIRE, WATER, GRASS, ELECTRIC, ICE, DRAGON -> 1f
                    ROCK, BUG, DARK -> 0.5f
                    else -> 0f
                }

                FLYING -> when (attackType) {
                    ROCK, ELECTRIC, ICE -> 2f
                    NORMAL, FLYING, POISON, GHOST, STEEL, FIRE, WATER, PSYCHIC, DRAGON, DARK, FAIRY -> 1f
                    FIGHTING, BUG, GRASS -> 0.5f
                    GROUND -> 0f
                    else -> 0f
                }

                POISON -> when (attackType) {
                    GROUND, PSYCHIC -> 2f
                    NORMAL, FLYING, ROCK, GHOST, STEEL, FIRE, WATER, ELECTRIC, ICE, DRAGON, DARK -> 1f
                    FIGHTING, POISON, BUG, GRASS, FAIRY -> 0.5f
                    else -> 0f
                }

                GROUND -> when (attackType) {
                    WATER, GRASS, ICE -> 2f
                    NORMAL, FIGHTING, FLYING, GROUND, BUG, GHOST, STEEL, FIRE, PSYCHIC, DRAGON, DARK, FAIRY -> 1f
                    POISON, ROCK -> 0.5f
                    ELECTRIC -> 0f
                    else -> 0f
                }

                ROCK -> when (attackType) {
                    FIGHTING, GROUND, STEEL, WATER, GRASS -> 2f
                    ROCK, BUG, GHOST, ELECTRIC, PSYCHIC, ICE, DRAGON, DARK, FAIRY -> 1f
                    NORMAL, FLYING, POISON, FIRE -> 0.5f
                    else -> 0f
                }

                BUG -> when (attackType) {
                    FLYING, ROCK, FIRE -> 2f
                    NORMAL, POISON, BUG, GHOST, STEEL, WATER, ELECTRIC, PSYCHIC, ICE, DRAGON, DARK, FAIRY -> 1f
                    FIGHTING, GROUND, GRASS -> 0.5f
                    else -> 0f
                }

                GHOST -> when (attackType) {
                    GHOST, DARK -> 2f
                    FLYING, GROUND, ROCK, STEEL, FIRE, WATER, GRASS, ELECTRIC, PSYCHIC, ICE, DRAGON, FAIRY -> 1f
                    POISON, BUG -> 0.5f
                    NORMAL, FIGHTING -> 0f
                    else -> 0f
                }

                STEEL -> when (attackType) {
                    FIGHTING, GROUND, FIRE -> 2f
                    GHOST, WATER, ELECTRIC, DARK -> 1f
                    NORMAL, FLYING, ROCK, BUG, STEEL, GRASS, PSYCHIC, ICE, DRAGON, FAIRY -> 0.5f
                    POISON -> 0f
                    else -> 0f
                }

                FIRE -> when (attackType) {
                    GROUND, ROCK, WATER -> 2f
                    NORMAL, FIGHTING, FLYING, POISON, GHOST, ELECTRIC, PSYCHIC, DRAGON, DARK -> 1f
                    BUG, STEEL, FIRE, GRASS, ICE, FAIRY -> 0.5f
                    else -> 0f
                }

                WATER -> when (attackType) {
                    GRASS, ELECTRIC -> 2f
                    NORMAL, FIGHTING, FLYING, POISON, GROUND, ROCK, BUG, GHOST, PSYCHIC, DRAGON, DARK, FAIRY -> 1f
                    STEEL, FIRE, WATER, ICE -> 0.5f
                    else -> 0f
                }

                GRASS -> when (attackType) {
                    FLYING, POISON, BUG, FIRE, ICE -> 2f
                    NORMAL, FIGHTING, ROCK, GHOST, STEEL, PSYCHIC, DRAGON, DARK, FAIRY -> 1f
                    GROUND, WATER, GRASS, ELECTRIC -> 0.5f
                    else -> 0f
                }

                ELECTRIC -> when (attackType) {
                    GROUND -> 2f
                    NORMAL, FIGHTING, POISON, ROCK, BUG, GHOST, FIRE, WATER, GRASS, PSYCHIC, ICE, DRAGON, DARK, FAIRY -> 1f
                    FLYING, STEEL, ELECTRIC -> 0.5f
                    else -> 0f
                }

                PSYCHIC -> when (attackType) {
                    BUG, GHOST, DARK -> 2f
                    NORMAL, FLYING, POISON, GROUND, ROCK, STEEL, FIRE, WATER, GRASS, ELECTRIC, ICE, DRAGON, FAIRY -> 1f
                    FIGHTING, PSYCHIC -> 0.5f
                    else -> 0f
                }

                ICE -> when (attackType) {
                    FIGHTING, ROCK, STEEL, FIRE -> 2f
                    NORMAL, FLYING, POISON, GROUND, BUG, GHOST, WATER, GRASS, ELECTRIC, PSYCHIC, DRAGON, DARK, FAIRY -> 1f
                    ICE -> 0.5f
                    else -> 0f
                }

                DRAGON -> when (attackType) {
                    ICE, DRAGON, FAIRY -> 2f
                    NORMAL, FIGHTING, FLYING, POISON, GROUND, ROCK, BUG, GHOST, STEEL, PSYCHIC, DARK -> 1f
                    FIRE, WATER, GRASS, ELECTRIC -> 0.5f
                    else -> 0f
                }

                DARK -> when (attackType) {
                    FIGHTING, BUG, FAIRY -> 2f
                    NORMAL, FLYING, POISON, GROUND, ROCK, STEEL, FIRE, WATER, GRASS, ELECTRIC, ICE, DRAGON -> 1f
                    GHOST, DARK -> 0.5f
                    PSYCHIC -> 0f
                    else -> 0f
                }

                FAIRY -> when (attackType) {
                    POISON, STEEL -> 2f
                    NORMAL, FLYING, GROUND, ROCK, GHOST, FIRE, WATER, GRASS, ELECTRIC, PSYCHIC, ICE, FAIRY -> 1f
                    FIGHTING, BUG, DARK -> 0.5f
                    DRAGON -> 0f
                    else -> 0f
                }

                STELLAR -> 1f
            }
        }

        fun getAttackTypeBoosterItemName(type: PokeType): String? {
            return when (type) {
                NORMAL -> "Silk Scarf"
                FIGHTING -> "Black Belt"
                FLYING -> "Sharp Beak"
                POISON -> "Poison Barb"
                GROUND -> "Soft Sand"
                ROCK -> "Hard Stone"
                BUG -> "Silver Powder"
                GHOST -> "Spell Tag"
                STEEL -> "Metal Coat"
                FIRE -> "Charcoal"
                WATER -> "Mystic Water"
                GRASS -> "Miracle Seed"
                ELECTRIC -> "Magnet"
                PSYCHIC -> "Twisted Spoon"
                ICE -> "Never-Melt Ice"
                DRAGON -> "Dragon Fang"
                DARK -> "Black Glasses"
                FAIRY -> "Fairy Feather"
                else -> null
            }
        }
    }
}